src/Datadog.AutoInstrumentation.NativeLoader/cor_profiler.cpp

```diff
@@ -0,0 +1,4587 @@
+            HRESULT hr = m_customProfiler->AppDomainCreationStarted(appDomainId);
+            if (FAILED(hr))
+            {
+                Warn("[Custom] Error in AppDomainCreationStarted() call.");
+                gHR = hr;
+            }
+        }
+
+        return gHR;
+    }
+
+    HRESULT STDMETHODCALLTYPE CorProfiler::AppDomainCreationFinished(AppDomainID appDomainId, HRESULT hrStatus)
+    {
+        HRESULT gHR = S_OK;
+
+        //
+        // Continuous Profiler
+        //
+        if (m_cpProfiler != nullptr)
+        {
+            HRESULT hr = m_cpProfiler->AppDomainCreationFinished(appDomainId, hrStatus);
+            if (FAILED(hr))
+            {
+                Warn("[Continuous Profiler] Error in AppDomainCreationFinished() call.");
+                gHR = hr;
+            }
+        }
+
+        //
+        // Tracer
+        //
+        if (m_tracerProfiler != nullptr)
+        {
+            HRESULT hr = m_tracerProfiler->AppDomainCreationFinished(appDomainId, hrStatus);
+            if (FAILED(hr))
+            {
+                Warn("[Tracer] Error in AppDomainCreationFinished() call.");
+                gHR = hr;
+            }
+        }
+
+        //
+        // Custom
+        //
+        if (m_customProfiler != nullptr)
+        {
+            HRESULT hr = m_customProfiler->AppDomainCreationFinished(appDomainId, hrStatus);
+            if (FAILED(hr))
+            {
+                Warn("[Custom] Error in AppDomainCreationFinished() call.");
+                gHR = hr;
+            }
+        }
+
+        return gHR;
+    }
+
+    HRESULT STDMETHODCALLTYPE CorProfiler::AppDomainShutdownStarted(AppDomainID appDomainId)
+    {
+        HRESULT gHR = S_OK;
+
+        //
+        // Continuous Profiler
+        //
+        if (m_cpProfiler != nullptr)
+        {
+            HRESULT hr = m_cpProfiler->AppDomainShutdownStarted(appDomainId);
+            if (FAILED(hr))
+            {
+                Warn("[Continuous Profiler] Error in AppDomainShutdownStarted() call.");
+                gHR = hr;
+            }
+        }
+
+        //
+        // Tracer
+        //
+        if (m_tracerProfiler != nullptr)
+        {
+            HRESULT hr = m_tracerProfiler->AppDomainShutdownStarted(appDomainId);
+            if (FAILED(hr))
+            {
+                Warn("[Tracer] Error in AppDomainShutdownStarted() call.");
+                gHR = hr;
+            }
+        }
+
+        //
+        // Custom
+        //
+        if (m_customProfiler != nullptr)
+        {
+            HRESULT hr = m_customProfiler->AppDomainShutdownStarted(appDomainId);
+            if (FAILED(hr))
+            {
+                Warn("[Custom] Error in AppDomainShutdownStarted() call.");
+                gHR = hr;
+            }
+        }
+
+        return gHR;
+    }
+
+    HRESULT STDMETHODCALLTYPE CorProfiler::AppDomainShutdownFinished(AppDomainID appDomainId, HRESULT hrStatus)
+    {
+        HRESULT gHR = S_OK;
+
+        //
```

Comment:
Would it make sense to use some kind of compile time macro  or template to make all this shorter?

Reply:
done in 4e9238c